presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,308 @@
+
+import javax.annotation.concurrent.Immutable;
+import javax.annotation.concurrent.NotThreadSafe;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class DynamicFilterService
+{
+    private final ConcurrentHashMap<SourceDescriptor, DynamicFilterSummaryWithSenders> dynamicFilterSummaries = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<SourceDescriptor, SettableFuture<DynamicFilterSummary>> futures = new ConcurrentHashMap<>();
+    private static final Logger log = Logger.get(DynamicFilterService.class);
+
+    public void storeOrMergeSummary(String queryId, String source, int stageId, int taskId, int driverId, DynamicFilterSummary dynamicFilterSummary, int expectedSummariesCount)
+    {
+        SourceDescriptor sourceDescriptor = SourceDescriptor.of(queryId, source);
+        DynamicFilterSummaryWithSenders dynamicFilterSummaryWithSenders = dynamicFilterSummaries.computeIfAbsent(sourceDescriptor,
+                ignored -> new DynamicFilterSummaryWithSenders());
+        DynamicFilterSummary mergedSummary = dynamicFilterSummaryWithSenders.addSummary(DynamicFilterSummaryWithSenders.StageTaskKey.of(stageId, taskId), driverId, dynamicFilterSummary, expectedSummariesCount);
+
+        if (mergedSummary != null) {
+            dynamicFilterSummaryWithSenders.getSummaryIfReady()
+                    .ifPresent(summary -> {
+                        SettableFuture<DynamicFilterSummary> future = futures.computeIfAbsent(sourceDescriptor, ignored -> SettableFuture.create());
+                        future.set(summary);
+                    });
+        }
+    }
+
+    /*
+     * registerTasks is required to know whether we have received dynamic filter summaries from all the tasks of query, after which the summary is ready to use
+     * Build side may start reporting summaries through storeOrMergeSummary before registerTasks runs
+     */
+    public void registerTasks(String source, Set<TaskId> taskIds)
+    {
+        if (taskIds.isEmpty()) {
+            return;
+        }
+
+        String queryId = taskIds.iterator().next().getQueryId().getId();
+        log.info("registerTasks source " + source + ", taskIds " + taskIds.stream().map(TaskId::toString).collect(Collectors.joining(",")));
+        checkArgument(taskIds.stream().allMatch(taskId -> taskId.getQueryId().getId().equals(queryId)), "All tasks have to belong to the same query");
```

Comment:
Here we are essentially checking that all elements of list `taskIds.stream().map(TaskId::getQueryId)` are the same, since we need to look at all elements of the list for this, I don't think it can be further improved.
I can change it to `taskIds.stream().map(TaskId::getQueryId).allMatch(isEqual(queryId))` to make it slightly less verbose.

Reply:
Sounds good.

> Here we are essentially checking that all elements of list `taskIds.stream().map(TaskId::getQueryId)` are the same, since we need to look at all elements of the list for this, I don't think it can be further improved.
> I can change it to `taskIds.stream().map(TaskId::getQueryId).allMatch(isEqual(queryId))` to make it slightly less verbose.

Sounds good.